### ISSUE TITLE
Tests are taking way too long

### ISSUE THREAD

**Issue by rosalinekarr** [filecoin-project/go-filecoin#2059]
### Description

The linux build is currently timing out as a result.

### Acceptance criteria

Tests should complete in less than 90 seconds to run.

**Comment by anorth** [comment]
Some of this is addressed in #1976, which reduce the total test time to about 3 minutes, I think. Daemon tests are the long pole here.

I agree there should be a shorter path opt-in too, probably skipping daemon tests.

**Comment by phritz** [comment]
we definitely cannot proceed with acceptance criteria as stated, just starting nodes and exchanging messages could easily take more than 10s. i would like there to be some analysis if we want to go down this route. my sense is that we should see where changes described in #2057 and #1976 get us before proceeding here.

we could also consider making use of testing.Short.

**Comment by phritz** [comment]
note https://github.com/filecoin-project/go-filecoin/issues/1976 is in progress

**Comment by rosalinekarr** [comment]
@phritz Updated from 10 seconds to 90 seconds.

**Comment by travisperson** [comment]
We have been seeing a timeout failure in our tests recently. They all timeout at the same time.

```
PASS
ok  	github.com/filecoin-project/go-filecoin/plumbing/mthdsig	0.050s
?   	github.com/filecoin-project/go-filecoin/plumbing/ntwk	[no test files]
?   	github.com/filecoin-project/go-filecoin/plumbing/ps	[no test files]
Too long with no output (exceeded 10m0s)
```

- https://circleci.com/gh/filecoin-project/go-filecoin/11369?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link
- https://circleci.com/gh/filecoin-project/go-filecoin/11377?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

This appears to be the first time it failed in master

- https://circleci.com/gh/filecoin-project/go-filecoin/11307

Might be related to: https://github.com/filecoin-project/go-filecoin/pull/1954 seems to be the PR when it started.

**Comment by travisperson** [comment]
I was able to reproduce this locally. At the output when the tests fail due to no output, my cpu was pegged to 100%. After a while, my cpu returned to normal levels, and no further output was produced. Test failed at the 30 minute timeout, and I was unable to determine which test was hanging. The go test output wasn't helpful here. I've since added some log messages in some tests I think might be causing issues.

**Comment by phritz** [comment]
commandeering this issue to be about the current p0 timeouts

**Comment by travisperson** [comment]
Note: I've seen this also after skipping the sector tests

**Comment by laser** [comment]
The failures are transient. I [retried @Kubuxu 's build a few times](https://circleci.com/gh/filecoin-project/go-filecoin/tree/feat%2Fstorage-error-leak) and it passed on the 4th try.

**Comment by phritz** [comment]
realtime conversation: https://filecoinproject.slack.com/archives/CEHHJNJS3/p1550874631208000

**Comment by travisperson** [comment]
- If `TestMessagePoolWait/waits` of _porcelain/mpool_test.go_ is skipped, I can repeatably run tests with no hanging. :1st_place_medal: 
- If I don't skip `TestMessagePoolWait/waits`, I see a majority of runs hang.
- If I don't skip `TestMessagePoolWait/waits`, but skip all tests of _porcelain/miner_test.go_, I see sometimes the test hang, but mostly it will pass.
- If I don't skip `TestMessagePoolWait/waits`, but skip all tests of _porcelain/miner_test.go_, **except** `TestMinerPreviewCreate`, I see the tests hang more often that with all porcelain miner tests skipped, at a pretty consistent rate.
- If I don't skip `TestMessagePoolWait/waits`, and only skip `TestMinerPreviewCreate`, I see about an equal number of passing and hanging runs.


Either of the following patches applied separately, result in no hanging.

```go
diff --git a/porcelain/mpool_test.go b/porcelain/mpool_test.go
index 6d2444aa..84ade872 100644
--- a/porcelain/mpool_test.go
+++ b/porcelain/mpool_test.go
@@ -76,6 +76,8 @@ func TestMessagePoolWait(t *testing.T) {
                        // The waiter must wait longer.
                        sub.Post(nil)
 
+                       time.Sleep(time.Second * 1)
+
                        // Add a message to the pool then bump the pubsub again.
                        plumbing.pending = types.NewSignedMsgs(1, signer)
                        sub.Post(nil)
diff --git a/proofs/sectorbuilder/testing/interface_test.go b/proofs/sectorbuilder/testing/interface_test.go
index 33c09ceb..ceefa1e3 100644
```

```go
diff --git a/plumbing/ps/testing.go b/plumbing/ps/testing.go
index 1af5af76..e6035b7c 100644
--- a/plumbing/ps/testing.go
+++ b/plumbing/ps/testing.go
@@ -20,7 +20,7 @@ type FakeSubscription struct {
 func NewFakeSubscription(topic string) *FakeSubscription {
        sub := &FakeSubscription{
                topic:       topic,
-               pending:     make(chan *pubsub.Message),
+               pending:     make(chan *pubsub.Message, 5),
                awaitCancel: sync.WaitGroup{},
        }
        sub.awaitCancel.Add(1)
diff --git a/proofs/sectorbuilder/testing/interface_test.go b/proofs/sectorbuilder/testing/interface_test.go
index 33c09ceb..ceefa1e3 100644
```

What appears to be happening is a golang scheduling issue. The issue appears to be that when a hang occurs, both `sub.Post(nil)` and `sub.pending =` are ran prior to the call to `subscription.Next` of `MessagePoolWait`. This means that on the first call to `subscription.Next`, the returned value of `plumbing.MessagePoolPending()` returns a pending list of `1`. `MessagePoolWait` then exists, but because the `pending` channel of `FakeSubscription` is not buffered, the second call to `sub.Post(nil)` hangs forever, blocking the wait group.

https://github.com/filecoin-project/go-filecoin/blob/b41f145f4c5ee6e8a1f16bdfb15a00ecbb081a9c/porcelain/mpool_test.go#L73-L85

https://github.com/filecoin-project/go-filecoin/blob/b41f145f4c5ee6e8a1f16bdfb15a00ecbb081a9c/porcelain/mpool.go#L20-L41

---------

This issue is resolved by the above two patches
1) The sleep forces the go runtime scheduler to schedule the other goroutine with the call to ` porcelain.MessagePoolWait`, this then esnures that the first call to  `plumbing.MessagePoolPending`, does not see the pending message from the assignment of `mpool_test.go#80`.
2) The buffering of the pending channel ensures that the second call to `sub.Post(nil)` will not block (note: size of 5 has no impact, it just needs to have room).

Either of the two patches above are not a good solution I believe, (the second isn't really a solution at all, and the timing is a scheduling hack). There is probably a better way to ensure the sequence of events would be to do the assignment to `pending` only after we can ensure the call to `plumbing.MessagePoolPending`.

Crude implementation: https://gist.github.com/travisperson/58535622ee84517fb4b5cb77495ef9f0

I've opened PR #2088 to skip the test till we have a good solution to make the test deterministic. 

**Testing Script**
```
# note: clean -testcache removes the go-filecoin binary, hence the build.
go clean -testcache
go build -o go-filecoin main.go
go test -v -parallel 8 ./...
```

**Comment by phritz** [comment]
@travisperson for president

**Comment by anorth** [comment]
It looks like that was entirely my fault, then 😞. Thanks for the state dump @travisperson I'll try to find a robust fix first thing Monday.

**Comment by ognots** [comment]
Great find on the cause of the hangs! 
It might be a bit premature to close this issue though. The acceptance criteria of this issue is not satisfied.
The linked issue https://github.com/filecoin-project/go-filecoin/issues/1976 is still open.
There are also still failing tests. For instance, this feature branch was created this morning from an updated `master` and fails after 5m29s.
Re-Opening to continue tracking testing optimizations.

**Comment by anorth** [comment]
Closure was based on phritz:
> commandeering this issue to be about the current p0 timeouts

I think we have other issues tracking everything else.

**Comment by phritz** [comment]
> It looks like that was entirely my fault

@anorth there is nobody here who hasn't written their fair share of concurrency bugs, don't sweat it!

**Comment by phritz** [comment]
yeah, think we should close this. if there are other/additional issues, let's have a new issue for it. 

re the big picture, i think as long as our tests take <10m to run we are in good shape
